File: python101/python2/27 Fájlok 10/Megoldások/valasztas.py
# Emelt szintű informatika érettségi, 2013. május

# 1. feladat ------------------------------------------------
print('1. feladat ------')
jeloltek = []
with open('szavazatok.txt','r') as fbe:
    for sor in fbe:
        s = sor.split()
        if s[4] == '-':
            s[4] = 'független'
        jeloltek.append([int(s[0]),int(s[1]),s[2],s[3],s[4]])
 
# 2. feladat ------------------------------------------------
print('2. feladat ------')
print('A helyhatósági választáson',
      len(jeloltek),'képviselőjelölt indult')

# 3. feladat ------------------------------------------------
print('3. feladat ------')
vnev = input('Vezetéknév: ')
knev = input('Keresztnév: ')
volt = False
for j in jeloltek:
    if j[2] == vnev and j[3] == knev:
        print('A jelölt',j[1],'szavazatot kapott.')
        volt = True
        break        
if not volt:
    print('Ilyen nevű képviselőjelölt nem szerepel',
          'a nyilvántartásban.')

# 4. feladat ------------------------------------------------
print('4. feladat ------')
VALASZTOK = 12345
szavazok = 0
for j in jeloltek:
    szavazok += j[1]
print('A választáson ',szavazok,' állampolgár,a jogosultak ',
      round(szavazok/VALASZTOK*100,2),'%-a vett részt.',sep='')

# 5. feladat ------------------------------------------------
print('5. feladat ------')
partok = {'GYEP'     :['Gyümölcsevők Pártja',0],
          'HEP'      :['Húsevők pártja',0],
          'TISZ'     :['Tejivók Szövetsége',0],
          'ZEP'      :['Zöldségevők Pártja',0],
          'független':['Független jelöltek',0]}
for j in jeloltek:
    partok[j[4]][1] += j[1]
for p in partok:
    print(partok[p][0],'= ',round(partok[p][1]/szavazok*100,2),
          '%',sep='')

# 6. feladat ------------------------------------------------
def szavazat(jelolt):
    return jelolt[1]

print('6. feladat ------')
legtobb = max(jeloltek,key=szavazat)
print('A legtöbb szavazatot kapták:')
for j in jeloltek:
    if j[1] == legtobb[1]:
        print(j[2],j[3],j[4])

# 7. feladat ------------------------------------------------
print('7. feladat ------')
kepviselok = {}
for j in jeloltek:
    if j[0] not in kepviselok or kepviselok[j[0]][0] < j[1]:
        kepviselok[j[0]] = j[1:]
with open('kepviselok.txt','w') as fki:
    for i in range(1,9):
        print(i,' '.join(kepviselok[i][1:]),file=fki)

File: java101/20190426 Valasztas/valasztas/src/valasztas/Valasztas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package valasztas;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

/**
 *
 * @author t1
 */
public class Valasztas {

    /**
     * @param args the command line arguments
     */
    
    static ArrayList<Jelolt> jeloltek = new ArrayList<>();
    static Map<String, Integer> partok = new HashMap<>();
    
    public static void main(String[] args) throws IOException {
        // 1.feladat
        Scanner be = new Scanner(new File("szavazatok.txt"));
        while (be.hasNextLine()) {            
            jeloltek.add(new Jelolt(be.nextLine()));
        }
        be.close();
        // 2.feladat
        // tmp.length vagy array.size - teljes méret
        System.out.println("2. feladat");
        System.out.println("A helyhatósági választásokra "
                + jeloltek.size() + "képviselőjelölt indult.");
        // 3. feladat
        System.out.println("3. feladat");
        Scanner bill = new Scanner(System.in);
        System.out.print("Vezeték és utónév: ");
        int sz = szavazat_keres(bill.nextLine());
        if (sz > -1 ) {
            System.out.println("Szavazatok száma: " + sz);
        } else {
            System.out.println("Ilyen nevű képviselőjelölt nem "
                    + "szerepel a nyilvántartásban");
        }
        // 4. feladat
        // összegzés tétel
        int osszes = 0;
        // összeadjuk az összes szavazatot, majd az egyes pártok szavazatait
        for (Jelolt j : jeloltek) {
            // csak a köetkező kell a 4-eshez
            osszes += j.getSzavazat();
            // ez kell az 5-öshöz
            Integer szdb =  partok.get(j.getPart());
            // megnézi hogy a hasmap-be van e már érték
            // nyilván nem lesz, ezért fel kell tölteni az elős értékkel a hasmap-et
            // nem lehet +=, csak = mert ha nincs azaz 0, ahhoz nem tudunk hozzáadni
            if (szdb == null) {
                szdb = j.getSzavazat();
            } else {
                szdb += j.getSzavazat();
            }
            partok.put(j.getPart(), szdb);
        }
        System.out.println("4. feladat");
        System.out.printf("A választáson %d állampolgár, "
                + " a jogosultak %.2f%%-a vett részt.\n" , 
                osszes, 100.0*osszes/12345);
        // 5. feladat
        System.out.println("5. feladat");
        System.out.println("A pártokra leadott szavazatok aránya:");
        for (String p : partok.keySet()) {
            System.out.printf("%s = %.2f%%\n", part_nev(p),
                    100.0*partok.get(p)/osszes);
        }
        
        // 6 .feladat
        // itt a listát rendezem
        System.out.println("6 .feladat");
        System.out.println("A legtöbb szavazatot katpák:");
        // csökkenő sorrendbe tettük a listát
        Collections.sort(jeloltek, (a,b) -> b.getSzavazat() - a.getSzavazat());
        int legtobb = jeloltek.get(0).getSzavazat();
        int i = 0;
        // listát addig  íratom ki ameddig nem egyenlő a legtöbbel 
        // 
        while (jeloltek.get(i).getSzavazat() == legtobb) {            
            System.out.println(jeloltek.get(i).getNev() + " "
            + jeloltek.get(i).getPart());
            i++;
        }
        
        // 7. feladat
        // listába lemegyek odáig ameddig nem találok 
        PrintWriter ki = new PrintWriter("kepviselok.txt");
        for (int kerulet = 1; kerulet < 8; kerulet++) {
            i = 0;
            while (jeloltek.get(i).getKerulet() != kerulet) {                
                i++;
            }
            ki.println(kerulet + " " + jeloltek.get(i).getNev()
                            + " " + jeloltek.get(i).getPart());
        }
        ki.close();
    }
    
    // lineráis keresés, ha nem találunk semmit, -1et adunk vissza
    private static int szavazat_keres(String nev) {
        for (int i = 0; i < jeloltek.size(); i++) {
            Jelolt j = jeloltek.get(i);
            if (j.getNev().equals(nev)) {
                return j.getSzavazat();
            }
        }
        return -1;
    }
    // case után nincs break, mert most return-al kilépünk
    private static String part_nev(String rov) {
        switch (rov) {
            case "GYEP":
                return "Gyümölcsevők pártja";
            case "HEP":
                return "Húsevők pártja";
            case "TISZ":
                return "Tejivók pártja";
            case "ZEP":
                return "Zöldségevők pártja";
            default:
                return "Függetlne jelöltek";
        }
    }
}
